### ISSUE TITLE
refinement ignored when applying abstraction?

### ISSUE THREAD

**Issue by ngrimm** [FStarLang/FStar#188]
I would expect the following code to typecheck:

```
module Test

val test : x:(option 'a) {is_Some x} -> Tot 'a
let test x = (fun y -> Some.v y) x
```

However, it fails with

```
fail5.fst(4,30-4,31): Subtyping check failed; expected type _122:(option ){(is_Some _122)}; got type (option )
```

**Comment by s-zanella** [comment]
Still reproducible.
```
$ bin/fstar.exe --universes examples/bug-reports/bug188.fst 
examples/bug-reports/bug188.fst(7,30-7,31): Subtyping check failed; expected type (uu___#5895:(Prims.option (?17596 x)){(Prims.b2t (Prims.is_Some uu___@0))}); got type (Prims.option (?17596 x))
```

**Comment by nikswamy** [comment]
I don't think I will fix this. 

What happens here is that `(fun y -> Some?.v y)` is first inferred to have a `PURE` type; then the rule for default effects of Pure functions kicks in---an unannotated function of type `(x:a -> PURE (b x) (wp x))` is checked to have type `x:a -> Tot (b x)`. In this case, this check fails, rightfully so, since the function is not unconditionally pure.

Fixing this program would require relaxing the default effect rule for `Tot` functions, which in most code we write, this default is the right thing to do.

In short, the rule is that an unannotated lambda term inferred to be in `PURE` is forced to be in `Tot`.
